luxonis_train/attached_modules/metrics/pml_metrics.py

```diff
@@ -0,0 +1,258 @@
+import torch
+from torch import Tensor
+
+from .base_metric import BaseMetric
+
+# Converted from https://omoindrot.github.io/triplet-loss#offline-and-online-triplet-mining
+# to PyTorch from TensorFlow
+
+
+def _pairwise_distances(embeddings, squared=False):
+    """Compute the 2D matrix of distances between all the embeddings.
+
+    @param embeddings: tensor of shape (batch_size, embed_dim)
+    @type embeddings: torch.Tensor
+    @param squared: If true, output is the pairwise squared euclidean
+        distance matrix. If false, output is the pairwise euclidean
+        distance matrix.
+    @type squared: bool
+    @return: pairwise_distances: tensor of shape (batch_size,
+        batch_size)
+    @rtype: torch.Tensor
+    """
+    # Get the dot product between all embeddings
+    # shape (batch_size, batch_size)
+    dot_product = torch.matmul(embeddings, embeddings.t())
+
+    # Get squared L2 norm for each embedding. We can just take the diagonal of `dot_product`.
+    # This also provides more numerical stability (the diagonal of the result will be exactly 0).
+    # shape (batch_size,)
+    square_norm = torch.diag(dot_product)
+
+    # Compute the pairwise distance matrix as we have:
+    # ||a - b||^2 = ||a||^2  - 2 <a, b> + ||b||^2
+    # shape (batch_size, batch_size)
+    distances = (
+        square_norm.unsqueeze(0) - 2.0 * dot_product + square_norm.unsqueeze(1)
+    )
+
+    # Because of computation errors, some distances might be negative so we put everything >= 0.0
+    distances = torch.max(distances, torch.tensor(0.0))
+
+    if not squared:
+        # Because the gradient of sqrt is infinite when distances == 0.0 (ex: on the diagonal)
+        # we need to add a small epsilon where distances == 0.0
+        mask = (distances == 0.0).float()
+        distances = distances + mask * 1e-16
+
+        distances = torch.sqrt(distances)
+
+        # Correct the epsilon added: set the distances on the mask to be exactly 0.0
+        distances = distances * (1.0 - mask)
+
+    return distances
+
+
+def _get_anchor_positive_triplet_mask(labels):
+    indices_equal = torch.eye(
+        labels.shape[0], dtype=torch.uint8, device=labels.device
+    )
+    indices_not_equal = ~indices_equal
+    labels_equal = labels.unsqueeze(0) == labels.unsqueeze(1)
+    mask = indices_not_equal & labels_equal
+    return mask
+
+
+class ClosestIsPositiveAccuracy(BaseMetric):
```

Comment:
Nitpick: Let's have classes on top and helper functions below. Or if they are used only inside specific class then make them class private functions?